Fix --program-suffix resulting in invalid symlink

wismill opened the pull request:
Please read [Github PR Conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#github-pull-request-conventions) and then fill in *one* of these two templates.

---

**Template Α: This PR modifies [behaviour or interface](https://github.com/cabalism/cabal/blob/master/CONTRIBUTING.md#changelog)**

Include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [x] The documentation has been updated, if necessary.
* [ ] [Manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) have been included.
* [ ] Tests have been added. (*Ask for help if you don’t know how to write them! Ask for an exemption if tests are too complex for too little coverage!*)

---

Fixes #9919
Fixes #8823

Based on [this comment](https://github.com/haskell/cabal/issues/8823#issuecomment-2079197342).

It took me a lot of time to understand where to look at and I am still unsure this fix does not introduce side effects. I would need some guidance if improvements are needed.

fendor commented on cabal-install/src/Distribution/Client/CmdInstall.hs:
```diff
@@ -540,7 +540,9 @@ installAction flags@NixStyleFlags{extraFlags, configFlags, installFlags, project
           traverseInstall (installCheckUnitExes InstallCheckInstall) installCfg
   where
-    configFlags' = disableTestsBenchsByDefault configFlags
+    -- Filter `configProgSuffix` out, in order to get the canonical executable
+    -- in the store.
+    configFlags' = (disableTestsBenchsByDefault configFlags){configProgSuffix = NoFlag}
```
Wouldn't this likely also apply to the flag `configProgPrefix`?

wismill replied:
Yes. I updated the code to fix it too.

fendor reviewed (commented):
Thanks for tackling this very annoying issue! I am wondering, whether it would be easy to add a regression test that makes sure this doesn't break again, e.g. a `PackageTests/ConfigProgTest` which installs a local executable with `--program-suffix` into the current directory and simply tests, whether the binary can be executed?

wismill commented:
I have not seen that somebody pushed some commits before force-pushing… Was it just a rebase?

wismill commented:
Updated code: fix `--program-prefix` as well, add a test for both affix options and both `--install-methods`. I think it fixes #8823 as well.

Reworked commit message for full explanation:

Currently the options `--program-{prefix,suffix}` for cabal install
affects the name of the file in the install directory *and* the
executable name in the store. The installation fails:

- If using `--install-method=symlink`, the *target* of the symlink is
  not affected by the affix options and it results in an invalid symlink.
- If using `--install-method=copy`, the copy fails because the source
  is not found.

Another issue is that it affects the computation of the hash of the 
build directory in the store, resulting in needless rebuild when using
successively different affix options.

Fixed by making the name of the executable in the store canonical, i.e.
always ignoring the program affix options.

Added a test for all the combinations of `--install-method` and program
affixes options.

wismill commented:
Rebased

fendor commented on cabal-install/src/Distribution/Client/CmdInstall.hs:
```diff
@@ -1101,6 +1103,14 @@ disableTestsBenchsByDefault configFlags =
     , configBenchmarks = Flag False <> configBenchmarks configFlags
     }
 
+-- | Disables program prefix and suffix.
```
Could we expand on this comment a bit? Like, why are we doing this, potentially linking to the issues, etc...
Otherwise, the reasoning might be lost in the future!

wismill replied:
Yes, I left the explanation at the use site, but since it is used only once let’s move it at the definition site. I expanded the explanation to explain both aspects: package hash & fix for install method.

fendor reviewed (approved):
LGTM, thanks! 

I have one documentation nitpick, please explain in `ignoreProgramAffixes` why we are ignoring these prefixes, so that the knowledge isn't lost in the future :)

ulysses4ever commented:
@mergify backport 3.12
